### ISSUE TITLE
Hide a large message

### ISSUE THREAD

**Issue by anhoangphuc** [ando-khachatryan/HiDDeN#9]
Hello. 
Firstly, sorry for my poor English. 
Your work is great. I'm very excited about digging into it. But I have a problem: I want to hide information in a (128x128) RGB image. Is it possible to embed a 3200bit length message  into the image in order to maintain bpp = 0.2 If it is not, what should I do?
Thank you!

**Comment by xytywh** [comment]
In my opinion,if you want to maintain bpp = 0.203 in a 128*128 RGB image,the message length should be 128*128*3*0.203 = 9978

**Comment by xytywh** [comment]
I have conducted a lot of experiments,such as 128 * 128 RGB,64 * 64 RGB,64 * 64 gray level image，32 * 32 gray level image，16 * 16 gray level image，all bpp is 0.203 ，but None of them can get the result like they describe in the paper,only for 16 * 16 gray level image and message length = 52，I can get 98% decode message accuracy。

**Comment by anhoangphuc** [comment]
If use such a huge message, I'm worried about overcomplicate model, when we have a Conv_net with in_channels is about 10,000, make our model overfit.

**Comment by xytywh** [comment]
You mean I can't hide 0.203 bpp message in a big image,such as 64 * 64 or even 128 * 128?

**Comment by anhoangphuc** [comment]
In addition, you can try to split (128 x 128 RGB) image into multiple (16 x 16 x 3) image, but I think it will make our encoded images less visual.

**Comment by xytywh** [comment]
I will try.

**Comment by yihanjiang** [comment]
I think this is an open issue of HiDDeN, since it copies the original message multiple times, which uses up the memory. My 1080 TI cannot handle 128 * 128, and the bit error rate is always around 1%.

To encode with large amount of information, some other network structure has to be tried?

**Comment by anhoangphuc** [comment]
Authors of this paper said that they compare this model to other methods in (512-512) cover image in BOSS. I don't understand how they do. :((.

**Comment by ando-khachatryan** [comment]
I have not tried any of the steganography experiments, only the watermarking. So take this with a grain of salt.
The problem with HiDDeN is not that it copies the message -- in fact in this implementation I have been careful not to. In encoder.py, 
```
    def forward(self, image, message):

        # First, add two dummy dimensions in the end of the message.
        # This is required for the .expand to work correctly
        expanded_message = message.unsqueeze(-1)
        expanded_message.unsqueeze_(-1)

        expanded_message = expanded_message.expand(-1,-1, self.H, self.W)
```
torch.expand() does not copy the tensor, it creates a "view".
The memory problem comes from Conv layer that "consumes" the expanded tensor - it has to have appropriate number of channels:
```
self.after_concat_layer = ConvBNRelu(self.conv_channels + 3 + config.message_length,
                                             self.conv_channels)
```
If you want to insert a 1000-bit message, you will need 64+3+1000 channels. But the memory/training speed is not the biggest issue: I don't think this would work properly for 1000-bit messages. The reason is that HiDDeN architecture, as I understand it, favors encoding the message "locally", aka the whole message is encoded into a small spatial window, and this is repeated over and over. With a very large message you cannot do this, you need to encode different chunks of the message into different spatial locations. 

@yihanjiang: There are other network structures which (supposedly) achieve much higher bpp for steganography, such as: https://arxiv.org/abs/1901.03892 and https://github.com/arnoweng/PyTorch-Deep-Image-Steganography. In my layman view, these approaches have some issues of their own, but are probably a better starting point if you want to encode huge message payloads.

**Comment by anhoangphuc** [comment]
@ando-khachatryan , I tried to encode 52bit-chunk of messages into different 16 x 16 area of image, but the encoded image looks quite bad. Do you have any idea to improve that?

**Comment by ando-khachatryan** [comment]
If the training loss is going down, then I'm not sure. 52 bits into 16x16 image is a lot.
